design/external-authorization-design.md

```diff
@@ -0,0 +1,669 @@
+2. The authorization systems can evolve at a different rate than Contour
+3. Authorization servers can be hosted in the Contour project by separate maintainers
+
+It also means that:
+
+1. Configuration is necessarily more complex and involves more components
+1. Some client authorization protocols may not be supported by Contour
+
+## Which Client Authorization Protocols to Support
+
+The Contour project should give operators practically useful authorization capabilities.
+This means that the project should build and support one or more authorization servers that will be useful to a wide audience.
+
+## Deployment Implications
+
+Adding authorization servers as a separate component increases the number of deployment options.
+Contour does not have a supported installer, and this would exacerbate that need.
+
+## GRPC Dependency
+
+This design proposes to only support the Envoy GRPC mechanism.
+This is a superset of the HTTP mechanism and allows more kinds of authorization to be implemented.
+The drawback is that many languages do not have good support for GRPC, so there is a smaller potential developer pool.
+Adding support for the HTTP mechanism would increase configuration complexity, but also introduce asymmetric feature support that depends on the auth server mechanism.
+There are few existing authorization servers that support the Envoy GRPC mechanism (most support HTTP).
+
+## No Authorization Intermediaries
+
+In contrast to [#1014](https://github.com/projectcontour/contour/pull/1014), Contour
+itself is not in the authorization path.
+This results in some restrictions (as noted above), and prevents
+the implementation of very fine-grained authorization (e.g different
+client authorization protocols per path).
+However, this approach means that the Contour changes are relatively limited.
+There is enough API that an external server can still remove any
+restrictions, but it will not feel like a native part of the Contour
+API.
+
+## Envoy JWT Filter
+
+As discussed earlier, the Envoy [JWT][6] filter can be used in conjunction with an OIDC authorization service.
+However, in general, an OIDC authorization server will need to implement JWT checking already, so the Envoy filter
+is really more of an optimization.
+Configuring the JWT filter adds more configuration complexity.
+Adding a JWT validation stanza to the `AuthorizationServer` struct would be possible, though arguably ugly.
+
+## SupportServiceReference kind field
+
+Contour doesn't need to support arbitrary types of support service resources.
+The design here includes a `Kind` field because it is modeled after `TypedLocalObjectReference`.
+
+## SupportService status
+
+One of the original requirements for adding dependencies on other services was being able to expose useful operational status.
+The `SupportService` status field is where this information would be exposed, but this design does not address what status should be exposed.
```

Comment:
I'd say at a minimum there should be a `Ready` condition (in this case indicating that the SupportService is ready to service requests). Probably, we'll want some fields eventually, but `Ready` allows for using the "wait for condition" flow easily.

Reply:
This can be dealt with later though; not blocking.